scripts/scil_compute_local_tracking.py

```diff
@@ -232,59 +339,67 @@ def main():
-        filtered_streamlines = (
-            compress_streamlines(s * voxres,
-                                 args.compress) / voxres
-            for s in filtered_streamlines)
-
-    tractogram = LazyTractogram(lambda: filtered_streamlines,
-                                data_per_streamlines,
-                                affine_to_rasmm=seed_img.affine)
-
-    filetype = nib.streamlines.detect_format(args.out_tractogram)
-    reference = get_reference_info(seed_img)
-    header = create_tractogram_header(filetype, *reference)
-
-    # Use generator to save the streamlines on-the-fly
-    nib.streamlines.save(tractogram, args.out_tractogram, header=header)
+    if not args.use_gpu:
+        # LocalTracking.maxlen is actually the maximum length
+        # per direction, so we need to divide by 2.0 to limit
+        # the streamline to the right length
+        max_steps_per_direction = int(args.max_length / 2.0 / args.step_size)
+        # make streamline slightly longer to filter out streamlines
+        # exceeding max_length
+        max_steps_per_direction += 1
+
+        streamlines_generator = LocalTracking(
+            _get_direction_getter(args),
+            BinaryStoppingCriterion(mask_data),
+            seeds, np.eye(4),
+            step_size=vox_step_size, max_cross=1,
+            maxlen=max_steps_per_direction,
+            fixedstep=True, return_all=True,
+            random_seed=args.seed,
+            save_seeds=True)
+
+    else:  # GPU tracking
+        max_strl_len = int(args.max_length / args.step_size) + 1
+        # make streamline slightly longer to filter out streamlines
+        # exceeding max_length
+        max_strl_len += 1
+
+        # data volume
+        odf_sh = odf_sh_img.get_fdata(dtype=np.float32)
+
+        streamlines_generator = GPUTacker(
+            odf_sh, mask_data, seeds,
+            vox_step_size, max_strl_len,
+            theta=get_theta(args.theta, args.algo),
+            sf_threshold=args.sf_threshold,
+            sh_interp=sh_interp,
+            sh_basis=args.sh_basis,
+            batch_size=batch_size,
+            forward_only=forward_only,
+            rng_seed=args.seed).track()
+
+    # filter streamlines by length on-the-fly
+    def filtered_strl_generator():
+        voxel_size = odf_sh_img.header.get_zooms()[0]
+        # this is the true max length, for filtering
+        # streamlines exceeding the max_length instead
+        # of cutting them
+        scaled_min_length = args.min_length / voxel_size
+        scaled_max_length = args.max_length / voxel_size
+        for strl, seed in streamlines_generator:
+            # filter by length
+            if scaled_min_length <= length(strl) <= scaled_max_length:
+                yield strl, seed
+
+    # use the filtered streamline generator to dump streamlines to file
+    _save_tractogram(filtered_strl_generator(),
+                     odf_sh_img, total_nb_seeds, args)
```

Comment:
The fact that you filter the streamlines before sending to _save_tractogram "breaks" the printing.

Master:
![image](https://github.com/scilus/scilpy/assets/4967417/44e89058-c588-4c0b-a671-2e75c6a3a9e3)

Your PR:
![image](https://github.com/scilus/scilpy/assets/4967417/32c3994e-1d04-499c-9213-fdbac9aeb7a9)

Reply:
Yes I don't know why it does that but I think it's because of the way the `LazyTractogram` class works (although that seems weird). I'll try and dig deeper.

Reply:
I couldn't fix the progress bar not reaching 100%, but I removed the first line. Also the progress bar now disappears once the tracking completes.